# Menyiapkan Library

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# Data Wrangling

## Gathering Data

### Load users_df

In [2]:
users_df = pd.read_csv('../data/users.csv')
users_df.head()

,student_id,name,birth_date,role,email,password,created_at,updated_at
0,S0001,Endah Gunawan,2014-11-07,siswa,endah.gunawan7522@gmail.com,gglfOR3Poo,2026-04-10 20:36:13,2026-04-24 06:00:12
1,S0002,Fitri Ginting,2014-07-02,siswa,fitri.ginting7171@gmail.com,8BRc5HE5vg,2026-04-20 03:52:08,2026-04-20 03:52:08
2,S0003,Hafiz Purnama,2014-04-27,siswa,hafiz.purnama7169@gmail.com,EDddm17eCd,2026-04-18 13:40:46,2026-04-24 16:39:25
3,S0004,Qori Wulandari,2013-10-19,siswa,qori.wulandari6789@gmail.com,fiJdGLJ3Pa,2026-02-14 06:15:12,2026-03-08 22:55:37
4,S0005,Oki Kusuma,2013-11-20,siswa,oki.kusuma6301@gmail.com,s7ZqmJVtwL,2026-03-02 22:09:36,2026-03-02 22:09:36


### Load chapters_df

In [3]:
chapters_df = pd.read_csv('../data/chapters.csv')
chapters_df.head()

,id,name,total_subchapter,created_at,updated_at
0,CT00001,Bilanga Bulat,3,2026-01-01 00:00:00,2026-01-01 00:00:00
1,CT00002,Bilangan Rasional,3,2026-01-01 00:00:00,2026-01-01 00:00:00
2,CT00003,Rasio,3,2026-01-01 00:00:00,2026-01-01 00:00:00
3,CT00004,Bentuk Aljabar,3,2026-01-01 00:00:00,2026-01-01 00:00:00
4,CT00005,Kesebangunan,3,2026-01-01 00:00:00,2026-01-01 00:00:00


### Load subchapters_df

In [4]:
subchapters_df = pd.read_csv('../data/subchapters.csv')
subchapters_df.head()

,id,chapter_id,name,created_at,updated_at
0,SCT00001,CT00001,Memahami Bilangan Bulat,2026-01-01 00:00:00,2026-01-01 00:00:00
1,SCT00002,CT00001,Operasi Hitung Bilangan Bulat,2026-01-01 00:00:00,2026-01-01 00:00:00
2,SCT00003,CT00001,Faktor Bilangan Bulat,2026-01-01 00:00:00,2026-01-01 00:00:00
3,SCT00004,CT00002,Di Antara Dua Bilangan Bulat,2026-01-01 00:00:00,2026-01-01 00:00:00
4,SCT00005,CT00002,Membandingkan Bilangan Rasional,2026-01-01 00:00:00,2026-01-01 00:00:00


### Load chapters_taken_df

In [5]:
chapters_taken_df = pd.read_csv('../data/chapters_taken.csv')
chapters_taken_df.head()

,id,user_id,chapter_id,created_at_x,updated_at,created_at
0,CHT00001,S0331,CT00002,2025-01-16 00:14:42,2026-04-27 23:18:52,2026-04-25 03:10:15
1,CHT00002,S0242,CT00002,2024-07-20 18:18:07,2026-04-06 12:52:29,2026-04-01 19:01:29
2,CHT00003,S0181,CT00006,2025-01-14 04:34:27,2026-04-12 00:35:07,2026-04-07 06:44:07
3,CHT00004,S0145,CT00003,2025-02-13 19:24:19,2026-04-27 23:18:52,2026-04-26 22:45:20
4,CHT00005,S0116,CT00006,2024-09-17 06:35:49,2026-04-20 18:32:43,2026-04-16 00:41:43


### Load assessments_df

In [6]:
assessments_df = pd.read_csv('../data/assessments.csv')
assessments_df.head()

,id,chapter_taken_id,subchapter_id,title,type,correct_answer,created_at,updated_at
0,AS00001,CHT01062,SCT00003,Assessment 1,exam,20.0,2026-04-19 06:00:27,2026-04-19 06:47:27
1,AS00002,CHT00314,SCT00009,Assessment 2,quiz,18.0,2026-04-27 16:47:57,2026-04-27 17:34:57
2,AS00003,CHT00935,SCT00007,Assessment 3,exam,24.0,2026-04-27 22:44:06,2026-04-27 23:18:52
3,AS00004,CHT00015,SCT00020,Assessment 4,placement,6.0,2026-04-25 13:55:29,2026-04-25 14:42:29
4,AS00005,CHT00800,SCT00004,Assessment 5,placement,19.0,2026-04-10 05:05:48,2026-04-10 05:52:48


### Load user_attempts_df

In [7]:
user_attempts_df = pd.read_csv('../data/user_attempts.csv')
user_attempts_df.head()

,id,user_id,assessment_id,score,completed_at,created_at,updated_at
0,UA00001,S9999,AS00193,34.80,2026-04-20 21:50:12,2026-04-20 21:08:12,2026-04-20 21:54:12
1,UA00002,S0180,AS00112,56.55,2026-04-25 09:26:43,2026-04-25 08:44:43,2026-04-25 09:30:43
2,UA00003,S0362,AS00056,86.55,2026-04-27 06:19:37,2026-04-27 05:37:37,2026-04-27 06:23:37
3,UA00004,S0702,AS00180,14.12,2026-04-27 23:18:52,2026-04-27 23:04:12,2026-04-27 23:18:52
4,UA00005,S0585,AS00153,8.11,2026-04-27 21:36:30,2026-04-27 20:54:30,2026-04-27 21:40:30


### Load user_progress_df

In [8]:
user_progress_df = pd.read_csv('../data/user_progress.csv')
user_progress_df.head()

,id,user_id,subchapter_id,chapter_taken_id,current_level,status,created_at,updated_at
0,UP00001,S0591,SCT00017,CHT00802,beginner,not started,2026-04-19 01:10:42,2026-04-20 01:27:42
1,UP00002,S0655,SCT00020,CHT00317,beginner,done,2026-04-22 08:32:30,2026-04-23 08:49:30
2,UP00002,S0655,SCT00020,CHT00317,beginner,done,2026-04-24 17:35:18,2026-04-25 17:52:18
3,UP00003,S0617,SCT00008,CHT00533,intermediate,not started,2026-04-25 18:53:39,2026-04-26 19:10:39
4,UP00004,S0966,SCT00001,CHT00638,intermediate,done,2026-04-19 00:35:00,2026-04-20 00:52:00


## Assessing Data

### Menilai data users_df

In [9]:
# Melihat struktur data
users_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   student_id  1000 non-null   object
 1   name        990 non-null    object
 2   birth_date  966 non-null    object
 3   role        978 non-null    object
 4   email       976 non-null    object
 5   password    982 non-null    object
 6   created_at  1000 non-null   object
 7   updated_at  1000 non-null   object
dtypes: object(8)
memory usage: 62.6+ KB


**Steps to Take:**
- Mengubah tipe data object pada kolom `birth_date`, `created_at` dan `updated_at` menjadi datetime
- Handling missing value dengan mempertimbangkan konteks data

**Insights**:
- Terdapat jumlah baris berbeda di setiap fitur. Hal ini menandakan bahwa `adanya missing value`.
- Tipe data pada fitur `birth_date`, `created_at`, dan `updated_at` masih `object`. Tipe data ini masih kurang tepat, harusnya datetime.
- Hal-hal ini akan ditangani di proses `cleaning data`

In [10]:
# Mengecek missing value
users_df.isna().sum()

student_id     0
name          10
birth_date    34
role          22
email         24
password      18
created_at     0
updated_at     0
dtype: int64

**Steps to Take:**
- Tidak ada

**Insights**:
- Sesuai dengan penjelasan sebelumnya bahwa terdapat `missing value` pada beberapa fitur.
- Hal ini akan diproses pada tahap `cleaning data`

In [11]:
# Mengecek jumlah duplikasi data
print(f'Jumlah duplikasi users_df: {users_df.duplicated().sum()}')

Jumlah duplikasi users_df: 0


**Steps to Take:**
- Tidak ada

**Insights**:
- `users_df` tidak memiliki data duplikat. Berarti ini tidak perlu dibersihkan pada tahap `cleaning data`

In [12]:
# Mengecek deskriptif statistik
users_df.describe(include='all')

,student_id,name,birth_date,role,email,password,created_at,updated_at
count,1000,990,966,978,976,982,1000,1000
unique,1000,561,720,6,976,975,1000,1000
top,S0001,???,2013-01-03,siswa,endah.gunawan7522@gmail.com,???,2026-04-10 20:36:13,2026-04-24 06:00:12
freq,1,13,5,952,1,8,1,1


**Steps to Take:**
- Menggunakan salah satu teknik handling missing value, pada tahap cleaning data

**Insights**:
- Karena banyak data kategorikal, maka tidak banyak informasi yang bisa didapatkan.
- Yang bisa didapatkan adalah fitur `name`, `birth_date`, `role`, `gmail`, dan `password` itu tidak memiliki `count` yang sama dengan fitur `student_id`. Hal ini menandakan adanya `missing value`

### Menilai data chapters_df

In [13]:
# Mengecek struktur data
chapters_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6 entries, 0 to 5
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   id                6 non-null      object
 1   name              6 non-null      object
 2   total_subchapter  6 non-null      int64 
 3   created_at        6 non-null      object
 4   updated_at        6 non-null      object
dtypes: int64(1), object(4)
memory usage: 372.0+ bytes


**Steps to Take:**
- Mengubah tipe data object pada kolom `created_at` dan `updated_at` menjadi datetime

**Insight**:
- Tidak ada `missing value`. Hal ini bisa dilihat pada jumlah baris di setiap fitur itu sama.
- Hanya saja ada tipe data yang kurang tepat pada fitur `created_at` dan `updated_at`. Seharusnya tipe datanya `datetime`
- Hal ini akan diproses pada tahap `cleaning data`

In [14]:
# Mengecek missing value
chapters_df.isna().sum()

id                  0
name                0
total_subchapter    0
created_at          0
updated_at          0
dtype: int64

**Steps to Take:**
- Tidak ada

**Insight**:
- Terbukti bahwa `chapters_df` tidak memiliki `missing_value`

In [15]:
# Mengecek jumlah duplikasi data
print(f'Jumlah duplikasi data chapters_df: {chapters_df.duplicated().sum()}')

Jumlah duplikasi data chapters_df: 0


**Steps to Take:**
- Tidak ada

**Insights**: 
- Tidak ada duplikasi data juga pada `chapters_df`

In [16]:
# Mengecek deskriptif statistik
chapters_df.describe(include='all')

,id,name,total_subchapter,created_at,updated_at
count,6,6,6.000000,6,6
unique,6,6,NaN,1,1
top,CT00001,Bilanga Bulat,NaN,2026-01-01 00:00:00,2026-01-01 00:00:00
freq,1,1,NaN,6,6
mean,NaN,NaN,3.500000,NaN,NaN
std,NaN,NaN,1.224745,NaN,NaN
min,NaN,NaN,3.000000,NaN,NaN
25%,NaN,NaN,3.000000,NaN,NaN
50%,NaN,NaN,3.000000,NaN,NaN
75%,NaN,NaN,3.000000,NaN,NaN


**Steps to Take:**
- Tidak ada

**Insights**:
- Pada fitur `total_subchapter`, terdapat `min = 3` dan `max = 6`
- Artinya, jumlah subbab yang paling minimal ialah 3 subbab dan paling banyak, 6 subbab.

### Menilai data subchapters_df

In [17]:
# Mengecek struktur data
subchapters_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21 entries, 0 to 20
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   id          21 non-null     object
 1   chapter_id  21 non-null     object
 2   name        21 non-null     object
 3   created_at  21 non-null     object
 4   updated_at  21 non-null     object
dtypes: object(5)
memory usage: 972.0+ bytes


**Steps to Take:**
- Mengubah tipe data object pada kolom `created_at` dan `updated_at` menjadi datetime

**Insights**:
- `subchapters_df` juga memiliki jumlah baris yang sama di setiap fitur.
- Artinya, ia `tidak memiliki missing value`.
- Tetapi pada fitur `created_at` dan `updated_at` memiliki tipe data `object`. Seharusnya `datetime`. Hal ini akan diproses pada tahap `cleaning_data`

In [18]:
# Mengecek missing value
subchapters_df.isna().sum()

id            0
chapter_id    0
name          0
created_at    0
updated_at    0
dtype: int64

**Steps to Take:**
- Tidak ada

**Insights**:
- Terbukti bahwa tidak ada missing value

In [19]:
# Mengecek jumlah duplikasi data
print(f'Jumlah duplikasi data subchapters_df: {subchapters_df.duplicated().sum()}')

Jumlah duplikasi data subchapters_df: 0


**Steps to Take:**
- Tidak ada

**Insights**:
- Tidak ada duplikasi data pada `subchapters_df`

In [20]:
# Mengecek deskriptif statistik
subchapters_df.describe(include='all')

,id,chapter_id,name,created_at,updated_at
count,21,21,21,21,21
unique,21,6,21,1,1
top,SCT00001,CT00006,Memahami Bilangan Bulat,2026-01-01 00:00:00,2026-01-01 00:00:00
freq,1,6,1,21,21


**Steps to Take:**
- Tidak ada

**Insights**:
- Pada baris count di setiap kolom, memiliki nilai yang sama. Artinya tidak ada indikasi missing value

### Menilai data chapters_taken_df

In [21]:
# Mengecek struktur data
chapters_taken_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1220 entries, 0 to 1219
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   id            1220 non-null   object
 1   user_id       1220 non-null   object
 2   chapter_id    1220 non-null   object
 3   created_at_x  1220 non-null   object
 4   updated_at    1220 non-null   object
 5   created_at    1220 non-null   object
dtypes: object(6)
memory usage: 57.3+ KB


**Steps to Take:**
- Menghapus kolom `created_at_x` pada tahap cleaning data
- Mengubah tipe data object pada kolom `created_at` dan `updated_at` menjadi datetime

**Insights**:
- Terdapat anomali fitur, yaitu `created_at_x`. Hal ini akan dihapus pada tahap `cleaning data`
- Pada fitur `created_at` dan `updated_at`, tipe datanya masih `object`. Pada tahap `cleaning data` akan diubah menjadi `datetime`
- Jumlah data di setiap fitur sama, artinya data ini `tidak memiliki missing value`

In [22]:
# Mengecek missing value
chapters_taken_df.isna().sum()

id              0
user_id         0
chapter_id      0
created_at_x    0
updated_at      0
created_at      0
dtype: int64

**Steps to Take:**
- Tidak ada

**Insights:**
- Terbukti bahwa `tidak ada missing value`

In [23]:
# Mengecek jumlah duplikasi data
print(f'Jumlah duplikasi data chapters_taken_df: {chapters_taken_df.duplicated().sum()}')

Jumlah duplikasi data chapters_taken_df: 0


**Steps to Take:**
- Tidak ada

**Insights**:
- Tidak ada jumlah duplikasi data

In [24]:
# Mengecek deskriptif statistik
chapters_taken_df.describe(include='all')

,id,user_id,chapter_id,created_at_x,updated_at,created_at
count,1220,1220,1220,1220,1220,1220
unique,1200,701,6,1200,949,1220
top,CHT00487,S0582,CT00004,2024-10-06 02:22:08,2026-04-27 23:18:52,2026-04-25 03:10:15
freq,2,8,216,2,272,1


**Steps to Take:**
- Tidak ada

**Insights**:
- Pada baris count di seluruh kolom, semua nya memiliki nilai yang sama, yaitu 1220.
- Artinya, semua kolom ini tidak ada indikasi missing value

### Menilai data assessments_df

In [25]:
# Mengecek struktur data
assessments_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206 entries, 0 to 205
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   id                206 non-null    object 
 1   chapter_taken_id  206 non-null    object 
 2   subchapter_id     206 non-null    object 
 3   title             206 non-null    object 
 4   type              206 non-null    object 
 5   correct_answer    195 non-null    float64
 6   created_at        206 non-null    object 
 7   updated_at        206 non-null    object 
dtypes: float64(1), object(7)
memory usage: 13.0+ KB


**Steps to Take:**
- Handle missing value pada kolom correct_answer
- Mengubah tipe data object pada kolom `created_at` dan `updated_at` menjadi datetime

**Insights:**
- Pada fitur `correct_answer`, ia memiliki jumlah baris berbeda dibanding dengan fitur lainnya
- Artinya, fitur itu memiliki `missing value`
- Pada tipe data `created_at` dan `updated_at`, tipe datanya akan diubah ke `datetime`

In [26]:
# Mengecek missing value
assessments_df.isna().sum()

id                   0
chapter_taken_id     0
subchapter_id        0
title                0
type                 0
correct_answer      11
created_at           0
updated_at           0
dtype: int64

**Steps to Take:**
- Handle missing value pada kolom correct_answer

**Insights:**
- Terbukti bahwa kolom correct_answer memiliki missing value sebanyak 11 data

In [27]:
# Mengecek jumlah duplikasi data
print(f'Jumlah duplikasi data assessments_df: {assessments_df.duplicated().sum()}')

Jumlah duplikasi data assessments_df: 0


**Steps to Take:**
- Tidak ada

**Insights:**
- Tidak ada duplikasi data pada data assessments_df

In [28]:
# Mengecek deskriptif statistik
assessments_df.describe(include='all')

,id,chapter_taken_id,subchapter_id,title,type,correct_answer,created_at,updated_at
count,206,206,206,206,206,195.000000,206,206
unique,200,186,21,200,4,NaN,206,201
top,AS00164,CHT00952,SCT00009,Assessment 164,placement,NaN,2026-04-19 06:00:27,2026-04-27 23:18:52
freq,2,2,17,2,74,NaN,1,6
mean,NaN,NaN,NaN,NaN,NaN,15.446154,NaN,NaN
std,NaN,NaN,NaN,NaN,NaN,6.179524,NaN,NaN
min,NaN,NaN,NaN,NaN,NaN,5.000000,NaN,NaN
25%,NaN,NaN,NaN,NaN,NaN,10.000000,NaN,NaN
50%,NaN,NaN,NaN,NaN,NaN,16.000000,NaN,NaN
75%,NaN,NaN,NaN,NaN,NaN,21.000000,NaN,NaN


**Steps to Take:**
- Tidak ada

**Insights:**
- Tidak ada

### Menilai data user_attempts_df

In [29]:
# Mengecek struktur data
user_attempts_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2053 entries, 0 to 2052
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             2053 non-null   object 
 1   user_id        2053 non-null   object 
 2   assessment_id  2053 non-null   object 
 3   score          2053 non-null   float64
 4   completed_at   2053 non-null   object 
 5   created_at     2053 non-null   object 
 6   updated_at     2053 non-null   object 
dtypes: float64(1), object(6)
memory usage: 112.4+ KB


**Steps to Take:**
- Mengubah tipe data object pada kolom `created_at` dan `updated_at` menjadi datetime 

**Insights:**
- Semua kolom memiliki jumlah baris yang sama. Artinya data user_attempts_df ini tidak memiliki indikasi missing value
- Tipe data untuk kolom `created_at` dan `updated_at` masih object. Seharusnya ini datetime. Hal ini akan ditangani di tahap cleaning data

In [30]:
# Mengecek missing value
user_attempts_df.isna().sum()

id               0
user_id          0
assessment_id    0
score            0
completed_at     0
created_at       0
updated_at       0
dtype: int64

**Steps to Take:**
- Tidak ada

**Insights:**
- Terbukti bahwa data user_attempts_df tidak memiliki missing value

In [31]:
# Mengecek jumlah duplikasi data
print(f'Jumlah duplikasi data user_attempts_df: {user_attempts_df.duplicated().sum()}')

Jumlah duplikasi data user_attempts_df: 0


**Steps to Take:**
- Tidak ada

**Insights:**
- Tidak ada duplikasi data

In [32]:
# Mengecek deskriptif statistik
user_attempts_df.describe(include='all')

,id,user_id,assessment_id,score,completed_at,created_at,updated_at
count,2053,2053,2053,2053.000000,2053,2053,2053
unique,2000,873,200,NaN,1927,2045,1919
top,UA01001,S0128,AS00169,NaN,2026-04-27 23:18:52,2026-04-27 23:18:14,2026-04-27 23:18:52
freq,2,8,24,NaN,124,2,132
mean,NaN,NaN,NaN,53.983176,NaN,NaN,NaN
std,NaN,NaN,NaN,75.231112,NaN,NaN,NaN
min,NaN,NaN,NaN,-50.000000,NaN,NaN,NaN
25%,NaN,NaN,NaN,25.420000,NaN,NaN,NaN
50%,NaN,NaN,NaN,49.020000,NaN,NaN,NaN
75%,NaN,NaN,NaN,73.790000,NaN,NaN,NaN


**Steps to Take:**
- Handling invalid value pada kolom `score`

**Insights:**
- Pada baris min kolom score, terdapat nilai -50. Hal ini perlu ditangani karena termasuk `invalid value`
- Pada baris max kolom score, terdapat nilai 999. Hal ini perlu ditangani karena termasuk `invalid value`
- Kedua hal ini akan ditangani pada tahap cleaning data

### Menilai data user_progress_df

In [33]:
# Mengecek struktur data
user_progress_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1552 entries, 0 to 1551
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   id                1552 non-null   object
 1   user_id           1552 non-null   object
 2   subchapter_id     1552 non-null   object
 3   chapter_taken_id  1552 non-null   object
 4   current_level     1476 non-null   object
 5   status            1552 non-null   object
 6   created_at        1552 non-null   object
 7   updated_at        1552 non-null   object
dtypes: object(8)
memory usage: 97.1+ KB


**Steps to Take:**
- Handling missing value pada kolom `current_level`
- Mengubah tipe data object pada kolom `created_at` dan `updated_at` menjadi datetime

**Insights:**
- Pada kolom `current_level` terdapat jumlah baris berbeda dengan kolom lainnya. Hal ini menandakan adanya indikasi missing value
- Pada kolom `created_at` dan `updated_at` masih memiliki tipe data object. Ini harus diubah menjadi datetime
- Semua hal di atas, akan dilakukan pada tahap cleaning data

In [34]:
# Mengecek missing value
user_progress_df.isna().sum()

id                   0
user_id              0
subchapter_id        0
chapter_taken_id     0
current_level       76
status               0
created_at           0
updated_at           0
dtype: int64

**Steps to Take:**
- Handling missing value pada kolom `current_level`

**Insights:**
- Terbukti bahwa pada kolom `current_level` terdapat missing value

In [35]:
# Mengecek jumlah duplikasi data
print(f'Jumlah duplikasi data user_progress_df: {user_progress_df.duplicated().sum()}')

Jumlah duplikasi data user_progress_df: 0


**Steps to Take:**
- Tidak ada

**Insights:**
- `user_progress_df` tidak memiliki duplikasi data

In [36]:
# Mengecek deskriptif statistik
user_progress_df.describe(include='all')

,id,user_id,subchapter_id,chapter_taken_id,current_level,status,created_at,updated_at
count,1552,1552,1552,1552,1476,1552,1552,1552
unique,1500,788,21,851,3,5,1551,1278
top,UP01500,S0364,SCT00008,CHT00952,advanced,in progress,2026-04-26 12:04:16,2026-04-27 23:18:52
freq,2,7,90,10,505,484,2,274


**Steps to Take:**
- Tidak ada

**Insights:**
- Pada baris count kolom current_level, ia memiliki jumlah baris yang berbeda sendiri. Ini sesuai insights sebelumnya, yaitu kolom ini memang memiliki missing value
- Pada baris top kolom current_level, terdapat nilai `advanced`. Hal ini menandakan bahwa kebanyakan siswa sudah diklasifikasikan skill matematikanya berada di level advanced
- Pada baris top kolom status, terdapat nilai `in progress`. Hal ini menandakan bahwa kebanyakan siswa masih banyak yang maish progress kelasnya dan belum selesai

## Cleaning Data

### Fixing users_df problem

In [37]:
# Mengecek struktur data
users_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   student_id  1000 non-null   object
 1   name        990 non-null    object
 2   birth_date  966 non-null    object
 3   role        978 non-null    object
 4   email       976 non-null    object
 5   password    982 non-null    object
 6   created_at  1000 non-null   object
 7   updated_at  1000 non-null   object
dtypes: object(8)
memory usage: 62.6+ KB


In [38]:
# Mendefinisikan kolom-kolom yang memiliki missing value
cols_with_missing = ['name', 'birth_date', 'role', 'email', 'password']

In [39]:
# Melihat baris yang NaN
users_df[users_df[['name', 'birth_date', 'role', 'email', 'password']].isnull().any(axis=1)]

,student_id,name,birth_date,role,email,password,created_at,updated_at
8,S0009,Anisa Siregar,2013-12-23,NaN,anisa.siregar2884@gmail.com,CheMcXnhjl,2026-03-20 06:20:18,2026-04-07 11:10:59
14,S0015,Santi Sari,2014-01-02,NaN,santi.sari3434@gmail.com,yODrPP5MNf,2026-04-15 19:54:44,2026-04-16 03:15:08
20,S0021,Prabowo Tarigan,NaN,siswa,prabowo.tarigan9551@gmail.com,VYVLkJNaHR,2026-02-28 04:44:19,2026-03-10 08:31:46
24,S0025,NaN,2013-08-25,siswa,gita.siregar5579@gmail.com,JrtN8cxBrG,2026-03-01 15:17:34,2026-03-22 11:32:27
32,S0033,Qomar Wulandari,2013-02-06,siswa,NaN,U13L0WEBaP,2026-03-03 02:12:49,2026-03-03 02:12:49
...,...,...,...,...,...,...,...,...
947,S0948,Rian Wijaya,NaN,siswa,rian.wijaya6017@gmail.com,UKYeuTssi0,2026-02-03 17:38:11,2026-03-18 11:14:16
966,S0967,Budi Hidayat,2012-01-30,siswa,NaN,K8nf4xGiIj,2026-03-06 23:59:30,2026-04-22 22:51:47
982,S0983,Siti Simanjuntak,NaN,siswa,siti.simanjuntak3340@gmail.com,31YQbGp4J9,2026-04-05 10:47:04,2026-04-05 10:47:04
985,S0986,Agus Wijaya,NaN,siswa,agus.wijaya8100@gmail.com,7O4S4XUeDy,2026-02-23 22:18:14,2026-03-03 23:25:43


**Insights:**
- Untuk kolom `name` yang kosong. Kita dapat mengisinya dari kolom `email` dan juga sebaliknya. Karena terdapat pola pengisian nama pada kolom `email`. Yaitu dengan `first_name.last_name@gmail.com`. Angkanya bisa kita abaikan untuk mengisian data 2 kolom tersebut.
- Pengisian kolom `name` atau `email`, kita akan menggunakan teknik yang namanya `imputation`.
- Unntuk kolom `role` terdapat yang NaN. Karena aplikasi kami dibuat hanya untuk siswa sejauh ini, maka nilai NaN ini bisa diisi dengan nilai `siswa`. Teknik yang digunakan adalah `imputation`
- Pada kolom `password`, untuk pengisian nilai NaN nya bisa dengan string acak sesuai dengan password milik user lain
- Untuk pengisian kolom `birth_date` yang memiliki nilai NaN, kami akan mengisinya pada tahun sekarang dikurangi umur anak kelas 7 SMP. Lalu tanggal dan bulannya akan kami acak. Teknik yang digunakan, `imputation`

#### Pengisian nilai NaN pada kolom Name dan Email

In [40]:
# Filter baris yang memiliki nilai null pada kolom email
filter_name_and_email_null = users_df['name'].notnull() & users_df['email'].isna()

In [41]:
# Melihat baris yang diinginkan
users_df[filter_name_and_email_null]

,student_id,name,birth_date,role,email,password,created_at,updated_at
32,S0033,Qomar Wulandari,2013-02-06,siswa,NaN,U13L0WEBaP,2026-03-03 02:12:49,2026-03-03 02:12:49
48,S0049,Rina Setiawan,2014-06-17,siswa,NaN,5kRkfBDXpr,2026-03-27 20:55:14,2026-04-26 14:47:28
89,S0090,Aditya Rahayu,2013-05-09,siswa,NaN,YUEw5FmE1J,2026-03-28 12:28:05,2026-03-28 12:28:05
161,S0162,Farhan Setiawan,2014-05-03,siswa,NaN,9rhEDwUGz2,2026-04-02 03:18:42,2026-04-14 02:41:25
178,S0179,Prabowo Utami,2013-11-10,siswa,NaN,w21uuBRx2A,2026-03-20 06:46:05,2026-03-25 22:37:45
181,S0182,Taufik Purnama,2013-07-28,siswa,NaN,3WEnQvhIHF,2026-03-13 07:38:41,2026-03-13 07:38:41
231,S0232,Endah Siregar,2014-07-13,siswa,NaN,3bDBryIeI3,2026-04-08 06:19:31,2026-04-08 06:19:31
319,S0320,Santi Situmorang,2013-04-03,siswa,NaN,sp7aX4WagX,2026-02-27 04:33:04,2026-03-19 02:36:35
413,S0414,Oki Harahap,2013-06-11,siswa,NaN,IxCvF20gdj,2026-02-26 22:23:15,2026-02-26 22:23:15
488,S0489,Dedi Kusuma,2013-02-16,siswa,NaN,QiT8JRLQTW,2026-04-01 11:23:45,2026-04-11 07:43:00


**Insights:**
- Di sini terdapat beberapa baris yang nilai nya NaN untuk kolom email.
- Kami akan mengisinya dengan mengambil first name dan last name pada kolom name. Pola pengisian nya yaitu `first_name.last_name@gmail.com`

In [42]:
# Mengisi nilai email yang NaN dengan pola yang sudah ditentukan
users_df.loc[filter_name_and_email_null, 'email'] = users_df.loc[filter_name_and_email_null, 'name'].str.lower().str.split().str.join('.') + '@gmail.com'
users_df[filter_name_and_email_null]

,student_id,name,birth_date,role,email,password,created_at,updated_at
32,S0033,Qomar Wulandari,2013-02-06,siswa,qomar.wulandari@gmail.com,U13L0WEBaP,2026-03-03 02:12:49,2026-03-03 02:12:49
48,S0049,Rina Setiawan,2014-06-17,siswa,rina.setiawan@gmail.com,5kRkfBDXpr,2026-03-27 20:55:14,2026-04-26 14:47:28
89,S0090,Aditya Rahayu,2013-05-09,siswa,aditya.rahayu@gmail.com,YUEw5FmE1J,2026-03-28 12:28:05,2026-03-28 12:28:05
161,S0162,Farhan Setiawan,2014-05-03,siswa,farhan.setiawan@gmail.com,9rhEDwUGz2,2026-04-02 03:18:42,2026-04-14 02:41:25
178,S0179,Prabowo Utami,2013-11-10,siswa,prabowo.utami@gmail.com,w21uuBRx2A,2026-03-20 06:46:05,2026-03-25 22:37:45
181,S0182,Taufik Purnama,2013-07-28,siswa,taufik.purnama@gmail.com,3WEnQvhIHF,2026-03-13 07:38:41,2026-03-13 07:38:41
231,S0232,Endah Siregar,2014-07-13,siswa,endah.siregar@gmail.com,3bDBryIeI3,2026-04-08 06:19:31,2026-04-08 06:19:31
319,S0320,Santi Situmorang,2013-04-03,siswa,santi.situmorang@gmail.com,sp7aX4WagX,2026-02-27 04:33:04,2026-03-19 02:36:35
413,S0414,Oki Harahap,2013-06-11,siswa,oki.harahap@gmail.com,IxCvF20gdj,2026-02-26 22:23:15,2026-02-26 22:23:15
488,S0489,Dedi Kusuma,2013-02-16,siswa,dedi.kusuma@gmail.com,QiT8JRLQTW,2026-04-01 11:23:45,2026-04-11 07:43:00


**Insights:**
- Kolom email yang NaN sudah terisi dengan pola `first_name.last_name@gmail.com`
- Selanjutnya kami akan melakukan hal berkebalikan. Yaitu mengisi kolom `name` yang NaN dengan mengambil first_name dan last_name dari kolom email

In [43]:
# Filter baris yang null pada kolom name
filter_name_null_and_email = users_df['name'].isna() & users_df['email'].notna()

In [44]:
# Melihat baris yang NaN pada kolom nameuse
users_df[filter_name_null_and_email]

,student_id,name,birth_date,role,email,password,created_at,updated_at
24,S0025,NaN,2013-08-25,siswa,gita.siregar5579@gmail.com,JrtN8cxBrG,2026-03-01 15:17:34,2026-03-22 11:32:27
54,S0055,NaN,2014-09-13,siswa,rian.pratama7564@gmail.com,rX1mQ8FKzO,2026-03-09 01:05:11,2026-03-09 01:05:11
76,S0077,NaN,2014-09-24,siswa,indra.permana4338@gmail.com,cBLsZhGlxp,2026-04-18 02:20:30,2026-04-20 21:16:07
133,S0134,NaN,2013-05-25,siswa,putri.ginting4428@gmail.com,JhZxBwksPj,2026-02-03 06:24:41,2026-04-05 08:05:47
430,S0431,NaN,2013-06-02,siswa,farhan.sembiring2094@gmail.com,iwAvVplipM,2026-03-11 12:09:27,2026-03-30 19:14:37
550,S0551,NaN,2015-11-02,siswa,lani.wijaya5909@gmail.com,CK1gvdCy74,2026-03-11 06:41:53,2026-03-11 06:41:53
724,S0725,NaN,2014-09-05,siswa,guntur.nasution4193@gmail.com,godd5WXWH7,2026-03-12 07:03:20,2026-03-19 22:20:55
741,S0742,NaN,2014-02-03,siswa,maya.nasution7642@gmail.com,T7FA2gQ9aB,2026-02-18 11:11:56,2026-02-18 11:11:56
750,S0751,NaN,2012-11-13,siswa,dewi.pratama4952@gmail.com,K1FXN5T6ij,2026-04-24 22:42:12,2026-04-26 03:37:53
911,S0912,NaN,2012-10-12,siswa,indah.hidayat7483@gmail.com,AswPUhKghB,2026-03-15 08:14:44,2026-03-15 08:14:44


**Insights:**
- Di sini terdapat beberapa baris yang nilai nya NaN untuk kolom name.
- Kami akan mengisinya dengan mengambil first name dan last name pada kolom email dan huruf pertama dari first_name dan last_name, akan dijadikan huruf besar lalu dipisahkan dengan spasi

In [49]:
# Mengambil first_name dan last_name dari email
first_name = users_df.loc[filter_name_null_and_email, 'email'].str.split(r'[.@]').str[0]
last_name = users_df.loc[filter_name_null_and_email, 'email'].str.split(r'[.@]').str[1].str.replace(r'\d+', '', regex=True)
first_name, last_name

(24       gita
 54       rian
 76      indra
 133     putri
 430    farhan
 550      lani
 724    guntur
 741      maya
 750      dewi
 911     indah
 Name: email, dtype: object,
 24       siregar
 54       pratama
 76       permana
 133      ginting
 430    sembiring
 550       wijaya
 724     nasution
 741     nasution
 750      pratama
 911      hidayat
 Name: email, dtype: object)

In [50]:
# Mengisi name dari kolom email yang sudah dipisah first dan last name nya
users_df.loc[filter_name_null_and_email, 'name'] = first_name.str.title() + ' ' + last_name.str.title()
users_df[filter_name_null_and_email]

,student_id,name,birth_date,role,email,password,created_at,updated_at
24,S0025,Gita Siregar,2013-08-25,siswa,gita.siregar5579@gmail.com,JrtN8cxBrG,2026-03-01 15:17:34,2026-03-22 11:32:27
54,S0055,Rian Pratama,2014-09-13,siswa,rian.pratama7564@gmail.com,rX1mQ8FKzO,2026-03-09 01:05:11,2026-03-09 01:05:11
76,S0077,Indra Permana,2014-09-24,siswa,indra.permana4338@gmail.com,cBLsZhGlxp,2026-04-18 02:20:30,2026-04-20 21:16:07
133,S0134,Putri Ginting,2013-05-25,siswa,putri.ginting4428@gmail.com,JhZxBwksPj,2026-02-03 06:24:41,2026-04-05 08:05:47
430,S0431,Farhan Sembiring,2013-06-02,siswa,farhan.sembiring2094@gmail.com,iwAvVplipM,2026-03-11 12:09:27,2026-03-30 19:14:37
550,S0551,Lani Wijaya,2015-11-02,siswa,lani.wijaya5909@gmail.com,CK1gvdCy74,2026-03-11 06:41:53,2026-03-11 06:41:53
724,S0725,Guntur Nasution,2014-09-05,siswa,guntur.nasution4193@gmail.com,godd5WXWH7,2026-03-12 07:03:20,2026-03-19 22:20:55
741,S0742,Maya Nasution,2014-02-03,siswa,maya.nasution7642@gmail.com,T7FA2gQ9aB,2026-02-18 11:11:56,2026-02-18 11:11:56
750,S0751,Dewi Pratama,2012-11-13,siswa,dewi.pratama4952@gmail.com,K1FXN5T6ij,2026-04-24 22:42:12,2026-04-26 03:37:53
911,S0912,Indah Hidayat,2012-10-12,siswa,indah.hidayat7483@gmail.com,AswPUhKghB,2026-03-15 08:14:44,2026-03-15 08:14:44


**Insights:**
- Kolom name sudah diisikan nilainya dan sudah sesuai

#### Pengisian nilai NaN pada kolom role

In [51]:
# Filter baris pada kolom role yang NaN
filter_role_null = users_df['role'].isna()

In [52]:
# Melihat baris yang memiliki nilai NaN pada kolom role
users_df[filter_role_null]

,student_id,name,birth_date,role,email,password,created_at,updated_at
8,S0009,Anisa Siregar,2013-12-23,NaN,anisa.siregar2884@gmail.com,CheMcXnhjl,2026-03-20 06:20:18,2026-04-07 11:10:59
14,S0015,Santi Sari,2014-01-02,NaN,santi.sari3434@gmail.com,yODrPP5MNf,2026-04-15 19:54:44,2026-04-16 03:15:08
126,S0127,Putri Wijaya,2013-12-23,NaN,putri.wijaya4750@gmail.com,S37tgi6v8z,2026-03-05 08:15:26,2026-04-26 00:19:59
192,S0193,Bambang Nasution,2014-07-26,NaN,bambang.nasution3249@gmail.com,gqJZUfh4or,2026-04-20 20:36:41,2026-04-24 14:36:29
197,S0198,Hany Wulandari,2014-04-12,NaN,hany.wulandari278@gmail.com,Gdqhf4kxrI,2026-02-14 00:19:00,2026-02-21 04:03:56
293,S0294,Fajar Hidayat,2014-10-17,NaN,fajar.hidayat9190@gmail.com,RZD9JOa9CT,2026-04-25 20:21:05,2026-04-25 20:21:05
386,S0387,Gilang Lestari,2013-05-13,NaN,gilang.lestari490@gmail.com,GTFndUytii,2026-04-15 10:08:33,2026-04-15 10:08:33
390,S0391,Gilang Setiawan,2014-08-24,NaN,gilang.setiawan5421@gmail.com,b1wQE5U17W,2026-04-27 13:40:58,2026-04-27 13:40:58
394,S0395,Ahmad Rahayu,2013-04-25,NaN,ahmad.rahayu2518@gmail.com,0sdOxKDJUn,2026-03-28 21:25:53,2026-04-24 15:12:00
474,S0475,Doni Hidayat,2013-02-27,NaN,doni.hidayat3828@gmail.com,Fw8LVoskuD,2026-03-10 16:20:48,2026-04-02 15:58:15


In [54]:
# Mengisi nilai NaN pada kolom role dengan teknik imputation
users_df['role'].fillna('siswa', inplace=True)

In [55]:
# Mengecek kembali
users_df[filter_role_null]

,student_id,name,birth_date,role,email,password,created_at,updated_at
8,S0009,Anisa Siregar,2013-12-23,siswa,anisa.siregar2884@gmail.com,CheMcXnhjl,2026-03-20 06:20:18,2026-04-07 11:10:59
14,S0015,Santi Sari,2014-01-02,siswa,santi.sari3434@gmail.com,yODrPP5MNf,2026-04-15 19:54:44,2026-04-16 03:15:08
126,S0127,Putri Wijaya,2013-12-23,siswa,putri.wijaya4750@gmail.com,S37tgi6v8z,2026-03-05 08:15:26,2026-04-26 00:19:59
192,S0193,Bambang Nasution,2014-07-26,siswa,bambang.nasution3249@gmail.com,gqJZUfh4or,2026-04-20 20:36:41,2026-04-24 14:36:29
197,S0198,Hany Wulandari,2014-04-12,siswa,hany.wulandari278@gmail.com,Gdqhf4kxrI,2026-02-14 00:19:00,2026-02-21 04:03:56
293,S0294,Fajar Hidayat,2014-10-17,siswa,fajar.hidayat9190@gmail.com,RZD9JOa9CT,2026-04-25 20:21:05,2026-04-25 20:21:05
386,S0387,Gilang Lestari,2013-05-13,siswa,gilang.lestari490@gmail.com,GTFndUytii,2026-04-15 10:08:33,2026-04-15 10:08:33
390,S0391,Gilang Setiawan,2014-08-24,siswa,gilang.setiawan5421@gmail.com,b1wQE5U17W,2026-04-27 13:40:58,2026-04-27 13:40:58
394,S0395,Ahmad Rahayu,2013-04-25,siswa,ahmad.rahayu2518@gmail.com,0sdOxKDJUn,2026-03-28 21:25:53,2026-04-24 15:12:00
474,S0475,Doni Hidayat,2013-02-27,siswa,doni.hidayat3828@gmail.com,Fw8LVoskuD,2026-03-10 16:20:48,2026-04-02 15:58:15


**Insights:**
- Sekarang kolom role yang NaN sudah terisi dengan siswa

#### Pengisian nilai NaN pada kolom password

# Exploratory Data Analysis